packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts

```diff
@@ -202,83 +308,118 @@ export default createRule({
+      typeNode: ts.Node,
+      checker: ts.TypeChecker,
+    ): TypeWithName[] {
+      if (ts.isParenthesizedTypeNode(typeNode)) {
+        return getUnionTypePart(typeNode.type, checker);
+      }
+      if (ts.isUnionTypeNode(typeNode)) {
+        return typeNode.types.flatMap(typeNode =>
+          getUnionTypePart(typeNode, checker),
+        );
+      }
+      const type = checker.getTypeAtLocation(typeNode);
       if (
-        typeNode.type === AST_NODE_TYPES.TSLiteralType &&
-        typeNode.literal.type === AST_NODE_TYPES.Literal
+        ts.isTypeReferenceNode(typeNode) &&
+        tsutils.isUnionOrIntersectionType(type)
       ) {
-        return [
-          {
-            typeFlags:
-              primitiveTypeFlagTypes[
-                typeof typeNode.literal
-                  .value as keyof typeof primitiveTypeFlagTypes
-              ],
-            typeName: describeLiteralTypeNode(typeNode),
-          },
-        ];
+        const node = getTypeNodeFromReferenceType(type);
+        if (node && node !== typeNode) {
+          return getUnionTypePart(node, checker);
+        }
       }
+      return [
+        {
+          type,
+          typeName: checker.typeToString(type),
+        },
+      ];
+    }
 
-      if (typeNode.type === AST_NODE_TYPES.TSUnionType) {
-        return typeNode.types.flatMap(getTypeNodeTypePartFlags);
+    function getInterSectionTypePart(
+      typeNode: ts.Node,
+      checker: ts.TypeChecker,
+    ): TypeWithName[] {
+      if (ts.isParenthesizedTypeNode(typeNode)) {
+        return getInterSectionTypePart(typeNode.type, checker);
       }
 
-      const nodeType = services.getTypeAtLocation(typeNode);
-      const typeParts = unionTypePartsUnlessBoolean(nodeType);
-
-      return typeParts.map(typePart => ({
-        typeFlags: typePart.flags,
-        typeName: describeLiteralType(typePart),
-      }));
-    }
-
-    function getTypeNodeTypePartFlagsCached(
-      typeNode: TSESTree.TypeNode,
-    ): TypeFlagsWithName[] {
-      const existing = typesCache.get(typeNode);
-      if (existing) {
-        return existing;
+      if (ts.isIntersectionTypeNode(typeNode)) {
+        return typeNode.types.flatMap(typeNode =>
+          getInterSectionTypePart(typeNode, checker),
+        );
       }
 
-      const created = getTypeNodeTypePartFlags(typeNode);
-      typesCache.set(typeNode, created);
-      return created;
+      const type = checker.getTypeAtLocation(typeNode);
+      if (
+        ts.isTypeReferenceNode(typeNode) &&
+        tsutils.isUnionOrIntersectionType(type)
+      ) {
+        const node = getTypeNodeFromReferenceType(type);
+        if (node && node !== typeNode) {
+          return getInterSectionTypePart(node, checker);
+        }
+      }
+      return [
+        {
+          type,
+          typeName: checker.typeToString(type),
+        },
+      ];
     }
 
     return {
       'TSIntersectionType:exit'(node: TSESTree.TSIntersectionType): void {
```

Comment:
[Refactor] The rule no longer cares about when it visits the node. We can simplify the visitors to not have `:exit`:

```suggestion
      TSIntersectionType(node: TSESTree.TSIntersectionType): void {
```

(this one and the other too)